### DISCUSSION TITLE
DG shape function pros and cons

### DISCUSSION THREAD

**Discussion by lindsayad** [libMesh/libmesh#3656] [Q&A]
Inspired by the question on https://github.com/idaholab/moose/pull/23986#discussion_r1344360255. What are pros and cons of `L2_LAGRANGE` vs `L2_HIERARCHIC` vs `MONOMIAL`? I am going to create a list here:

### L2_LAGRANGE

Pros:
- Implementation simplicity

Cons:
- limited to a max of a third order basis

### L2_HIERARCHIC

Pros:
- Arbitrarily high order

Cons:

### MONOMIAL

Pros:
- Arbitrarily high order
- less global dofs for meshes with non-simplicial elements

Cons:
- Not as good of an error constant when cross-terms are dropped, e.g. for a P=1 basis you don't get the bilinear term xy that you do with LAGRANGE/HIERARCHIC
- Worse condition number, particularly at high polynomial order

**Comment by jwpeterson** [comment]
I think that `L2_HIERARCHIC` are going to have better conditioning at high order than `MONOMIAL`. Not sure about `L2_LAGRANGE` in that regard.

**Reply by dknez** [reply]
I agree, `MONOMIAL` is pretty much "worst case" for condition number IIRC. If your order isn't too high, then that doesn't matter much, though.

**Comment by dknez** [comment]
The reason I added `L2_LAGRANGE` is to enable "nodal DG methods". There is a textbook from Hesthaven about this approach, for example. The main (only?) advantage of this compared to other approaches is simplicity of implementation.

**Reply by jwpeterson** [reply]
Yeah, computing the `nodal_soln` for an `Elem` is trivially easy for `L2_LAGRANGE`, for instance. We use them for stress postprocessing.

**Comment by roystgnr** [comment]
`MONOMIAL` has the same number of DoFs (for the same p level) on simplicial elements, but fewer DoFs on any mesh that has quads/prisms/pyramids/hexes.

It gets that by dropping the "cross terms" - e.g. for dim=2 p=2 you get x², y², xy, but you don't get x²y or y²x or x²y².

In CG problems people generally think those cross terms are worth it - you get a better convergence constant and slightly faster integration and the cost is slightly more bandwidth in your sparsity pattern - but it's not as obvious in DG.  I'd actually be somewhat interested to see experiments.  If it was worthwhile, it'd be possible to create a `FEFamily` that covers the same polynomial space as `MONOMIAL` but has a better-conditioned basis.

**Comment by lindsayad** [comment]
Is there any equivalence between (L2_)HIERARCHIC and (L2_)LAGRANGE for given element types/polynomial order? Just glancing cursorily at the source code it looks like there is equivalence in 1D for first order ... but that's about it...?

**Reply by roystgnr** [reply]
The function spaces are the same at any order, but the bases are different for p>1 on any element type.

**Comment by jwpeterson** [comment]
@lindsayad you may want to add the `FEXYZ` to your comparison list as well. It is basically the same thing as `MONOMIAL` but the `phi` values depend on each element's physical coords, and there is some benefit to this extra work, though I can't remember exactly what at the moment.

**Reply by lindsayad** [reply]
Can maintainers edit others' posts? ... I guess I can answer my own question on that: yes
